### PR TITLE
samples: drivers: added sample app for LCD 2004

### PR DESCRIPTION
**Pull request by andrei-edward-popa**
added sample application for LCD 2004 through I2C
display text on LCD 2004

### CODE REVIEW

**Review by gmarull — CHANGES_REQUESTED** [review]
I think we should create a new API for character-based displays. There's actually another device _supported_ today via a custom API (see https://github.com/zephyrproject-rtos/zephyr/pull/42237), so it's probably time to find a common ground since these type of devices are fairly common.

**Comment by tejlmand** [comment]
> I think we should create a new API for character-based displays.

Sounds like a good plan. @gmarull any idea on the time frame for such work ?
Should we consider to let PRs like this go in until such work has completed, or will that just add up to a lot of re-work later you think.

**Comment by gmarull** [comment]
> > I think we should create a new API for character-based displays.
> 
> Sounds like a good plan. @gmarull any idea on the time frame for such work ? Should we consider to let PRs like this go in until such work has completed, or will that just add up to a lot of re-work later you think.

I think we could consider adding a driver to `misc/`, same as I suggested for https://github.com/zephyrproject-rtos/zephyr/pull/42237. But I think one of the premises of `misc` drivers is that they are sort of “unique”, i.e., creating an abstraction would not make much sense. I think this device is not the case: there are many character-based displays.

**Comment by carlescufi** [comment]
API meeting

- Consensus to add a new character display API for hardware that is designed to display only characters instead of a device-specific sample, since we already have one (grove_lcd_rgb)

**Comment by de-nordic** [comment]
> API meeting
> 
>     * Consensus to add a new character display API for hardware that is designed to display only characters instead of a device-specific sample, since we already have one (grove_lcd_rgb)

The LCD2004 is actually HD44780  based two liner with NXP PCF8574T i2c 8 bit i/o expander.

We already have the HD44780 driver, as sample, https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/lcd_hd44780/

We probably need  the PCF8574T driver and DTS bindings, as this is i2c device and a way to attach the HD44780 directly to GPIO, as a parallel port controlled device, or as i2c device.

Seems to get complicated.

Documents:
https://www.sparkfun.com/datasheets/LCD/HD44780.pdf
https://www.nxp.com/docs/en/data-sheet/PCF8574_PCF8574A.pdf

File: samples/drivers/i2c_lcd_2004/src/main.c [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+static inline inline void lcd_char(const struct device *dev, char ch)
+{
+	lcd_send_byte(dev, ch, LCD_CHARACTER);
+}
+
+static inline void lcd_string(const struct device *dev, const char *str)
+{
+	for (uint32_t i = 0; i < strlen(str); i++) {
+		lcd_char(dev, str[i]);
+	}
```
Comment (kemonats) [inline samples/drivers/i2c_lcd_2004/src/main.c:89]: Loop without `strlen()` will be better e.g:
```suggestion
	while (*str != '\0') {
		lcd_char(dev, *str);
		str++
	}
```

Comment (andrei-edward-popa) [reply samples/drivers/i2c_lcd_2004/src/main.c]: Done!

**Review by carlescufi — APPROVED** [review]
LGTM on first look

**Review by gmarull — CHANGES_REQUESTED** [review]
As I mentioned earlier, I think we should not add more character LCD samples/drivers, but instead come up with an API. We already have, e.g., the grove_lcd misc driver that could benefit from it.

**Pull request closed without merging**